MapboxVisionTests/Mocks/MockVideoSource.swift

```diff
@@ -1,10 +1,101 @@
 import Foundation
+
 @testable import MapboxVision
 
-class MockVideoSource: VideoSource {
-    var isExternal: Bool = true
+final class MockVideoSource: ObservableVideoSource {
+    private enum Constants {
+        static let fps: Int32 = 30
+        static let frameDelay = 1.0 / Double(fps)
+        static let frameWidth = 960
+        static let frameHeight = 540
+        static let bitsInByte = 8
+    }
+
+    private let processingQueue = DispatchQueue(label: "com.mapbox.RecordCoordinatorTest.VideoSource")
+    private lazy var timer: DispatchSourceTimer = {
+        let timer = DispatchSource.makeTimerSource()
+        timer.schedule(deadline: .now() + Constants.frameDelay, repeating: Constants.frameDelay)
+        timer.setEventHandler { [weak self] in
+            self?.emitNewBuffer()
+        }
+        return timer
+    }()
+    private var framesAlreadyWritten: Int64 = 0
+
+    override init() {
+        super.init()
+    }
+
+    func start() {
+        timer.resume()
+    }
+
+    func stop() {
+        timer.suspend()
+    }
+
+    private func emitNewBuffer() {
+        var buffer: CMSampleBuffer?
+
+        guard let pixelBuffer = makePixelBuffer() else {
+            return
+        }
+
+        let duration = CMTime(value: 1, timescale: Constants.fps)
+        let presentationTimeStamp = CMTime(value: framesAlreadyWritten, timescale: Constants.fps)
+        var sampleTimingInfo = CMSampleTimingInfo(duration: duration, presentationTimeStamp: presentationTimeStamp, decodeTimeStamp: CMTime.invalid)
+        CMSampleBufferCreateReadyWithImageBuffer(allocator: kCFAllocatorDefault,
+                                                 imageBuffer: pixelBuffer,
+                                                 formatDescription: makeFormatDescription(from: pixelBuffer)!,
+                                                 sampleTiming: &sampleTimingInfo,
+                                                 sampleBufferOut: &buffer)
+
+        framesAlreadyWritten += 1
+        let sample = VideoSample(buffer: buffer!, format: Image.Format.BGRA)
+
+        self.notify { observer in
+            observer.videoSource(self, didOutput: sample)
+        }
+    }
+
+    private func makePixelBuffer() -> CVPixelBuffer? {
+        let attrs = [
+            kCVPixelBufferMetalCompatibilityKey: kCFBooleanTrue
+        ] as CFDictionary
+        var pixelBufferUnwrapped: CVPixelBuffer?
+        let status = CVPixelBufferCreate(kCFAllocatorDefault, Int(Constants.frameWidth), Int(Constants.frameWidth), kCVPixelFormatType_32BGRA, attrs, &pixelBufferUnwrapped)
+        guard status == kCVReturnSuccess, let pixelBuffer = pixelBufferUnwrapped else {
+            return nil
```

Comment:
Can we have a situation where `status != kCVReturnSuccess` AND `pixelBuffer != nil`?
Should we perform releasing for memory then?

Reply:
it is possible, yes. Releasing what do you propose?

Reply:
Oh, I'm sorry, ARC does the job. There's no need for additional operations.